internal/common/commontest/inputs.go

```diff
@@ -0,0 +1,64 @@
+package commontest
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+)
+
+// AddUserInput adds user controllable data to the request r.
+// The data type can be configured with inputType. If inputType is not
+// supported, the program exits.
+// You can also specify the key and value of the data to be added to
+// the request. The key "input" and value "fake-user-input"
+// are used by default.
+func AddUserInput(inputType string, r *http.Request, key, value string) {
+	if key == "" {
+		key = "input"
+	}
+	if value == "" {
+		value = "fake-user-input"
+	}
+	switch inputType {
+	case "query", "buffered-query":
+		q := r.URL.Query()
+		q.Add(key, value)
+		r.URL.RawQuery = q.Encode()
+	case "body", "buffered-body":
+		v := make(url.Values)
+		v.Set(key, value)
+		r.Body = io.NopCloser(strings.NewReader(v.Encode()))
+	case "cookies":
+		r.AddCookie(&http.Cookie{
+			Name:  key,
+			Value: value,
+		})
+	case "headers":
+		r.Header.Set(key, value)
+	case "params":
+		r.URL.Path = path.Join(r.URL.Path, value)
+	case "response":
+		// BUG: This endpoint doesn't actually read a response.
+		// For now, just add a header since it's quick
+		r.Header.Set(key, value)
+	default:
+		log.Fatalf("unknown input type: %s", inputType)
+	}
+}
+
+func addHeadersJSONCreds(r *http.Request) {
+	var creds struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}
+	creds.Username = "Robert'; DROP TABLE Students;--"
+	data, err := json.Marshal(creds)
+	if err != nil {
+		log.Fatalf("failed to marshal JSON object: %s", err)
+	}
+	r.Header.Set("credentials", string(data))
```

Comment:
nit: I know this is unlikely to happen but it looks like `r.Header` can panic here if `r` is nil.

Reply:
That's true. I don't think we can recover from being passed a nil `*http.Request` though, so I'm not sure if there's any action to take here. We could `log.Fatal` but I think the stack trace of the panic might be more useful for debugging.